build/result.go

```
@@ -22,166 +20,258 @@ import (
-			})
+			return res, err
+		}, ch)
+		select {
+		case _, ok := <-done:
+			if !ok {
+				cancel()
+				return
+			}
+		default:
+		}
+		close(baseCh)
+
+		if resp != nil {
+			respCh <- resp
+		}
+		if err != nil {
+			resCtxCh <- nil
+			errCh <- err
+			cancel()
+			return
+		}
 
-			// Record the client and ctx as well so that containers can be created from the SolveError.
-			resultCtx.res = res2
-			resultCtx.gwClient = c
-			resultCtx.gwCtx = ctx
-			resultCtx.gwDone = func() {
-				cancel(doneErr)
-				// wait for Build() completion(or timeout) to ensure the Build's finalizing and avoiding an error "context canceled"
+		go func() {
+			def, err := getDefinition(ctx, res)
+			if err != nil {
+				resCtxCh <- nil
+				errCh <- err
+				cancel()
+				return
+			}
+
+			// NOTE: ideally this second connection should be lazily opened
+			// NOTE: or... we could open this session *during* the other
+			//       session, which would let us grab the same vertices
+			//       immediately, mitigating the risk for "inconsistent solve
+			//       result"
+			opt := opt
+			opt.Ref = ""
+			_, err = cc.Build(ctx, opt, "buildx", func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+				res, err := evalDefinition(ctx, c, def)
+				if err != nil {
+					return nil, errors.Wrap(err, "inconsistent solve result")
+				}
+				resCtx := &ResultContext{
+					done:     make(chan struct{}),
+					res:      res,
+					gwClient: c,
+					gwCtx:    ctx,
+				}
+				close(done)
+				resCtxCh <- resCtx
 				select {
-				case <-buildDoneCh:
-				case <-time.After(5 * time.Second):
+				case <-resCtx.done:
+					return nil, nil
+				case <-ctx.Done():
+					return nil, ctx.Err()
 				}
-			}
+			}, nil)
 			select {
-			case resultCtxCh <- &resultCtx:
-			case <-ctx.Done():
-				return nil, ctx.Err()
+			case _, ok := <-done:
+				if !ok {
+					cancel()
+					return
+				}
+			default:
 			}
+			if err != nil {
+				resCtxCh <- nil
+				errCh <- err
```

At this point, both of `respCh` and `resCtxCh` have already been sent so the the for loop receiving `errCh`  isn't polling `errCh` anymore. So maybe this `errCh <- err` seems to block infinitly?

---

Good point, I think we can discard the error in this case.